backupccl: ignore all dropped descriptors during backup

adityamaru opened the pull request:
Previously, descriptors that were resolved at `EndTime`
and were in a `DROP` state were not treated uniformly. While
we ignored table descriptors, we continued to backup database,
schema and type descrpitors.

This resulted in  atleast two bugs:
1) If a database descriptor was in a dropped state, and a new
descriptor with the same name was created then a BACKUP DATABASE
of the new database would fail during resolution.

2) A cluster backup of the above state would succeed, but since it
would include duplicate entries for the same name database, the restore
of such a backup would fail.

This change unifies the behaviour by ignoring all DROP descriptors seen
by the backup at `EndTime`. A follow up PR will teach restore to ignore
all dropped descriptors so as to allow users with "corrupt" backups as
explained in 2) to be able to restore.

Informs: #76517

Release note (bug fix): Backup incorrectly backed up database, schema,
and type descriptors that were in a DROP state at the time the backup was
run. This bug resulted in the user being unable to backup and restore if
their cluster had dropped and public descriptors with colliding names.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76635)
<!-- Reviewable:end -->

adityamaru committed:
backupccl: ignore all dropped descriptors during backup

Previously, descriptors that were resolved at `EndTime`
and were in a `DROP` state were not treated uniformly. While
we ignored table descriptors, we continued to backup database,
schema and type descrpitors.

This resulted in  atleast two bugs:
1) If a database descriptor was in a dropped state, and a new
descriptor with the same name was created then a BACKUP DATABASE
of the new database would fail during resolution.

2) A cluster backup of the above state would succeed, but since it
would include duplicate entries for the same name database, the restore
of such a backup would fail.

This change unifies the behaviour by ignoring all DROP descriptors seen
by the backup at `EndTime`. A follow up PR will teach restore to ignore
all dropped descriptors so as to allow users with "corrupt" backups as
explained in 2) to be able to restore.

Release note (bug fix): Backup incorrectly backed up database, schema,
and type descriptors that were in a DROP state at the time the backup was
run. This bug resulted in the user being unable to backup and restore if
their cluster had dropped and public descriptors with colliding names.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76635#-:-Mw70mgE9O6yklGl1gDw:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @msbutler)

___
*[pkg/ccl/backupccl/backupresolver/targets.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76635#-Mw70_PC2MmRyeUh2wo-:-Mw70_PC2MmRyeUh2wo0:b-6banij) ([raw file](https://github.com/cockroachdb/cockroach/blob/451274be01d42920d3e85f5e1a828998ded4d54a/pkg/ccl/backupccl/backupresolver/targets.go#L341)):*
> ```Go
> 		if _, ok := alreadyRequestedDBs[dbID]; !ok {
> 			desc := r.DescByID[dbID]
> 			// Verify that the database is in the correct state.
> ```

Note that we do this rather than just filtering because we already resolved this database by ID above?



_Code quote:_
```Go
te.
```


<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
LGTM

adityamaru commented:
blathers backport 21.2
